Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -534,6 +516,49 @@ public class BrowserViewController: UIViewController, BrowserViewControllerDeleg
     try? widgetBookmarksFRC?.performFetch()
 
     updateWidgetFavoritesData()
+    
+    // Lists need to be compiled before attempting tab restoration
+    Task(priority: .high) {
+      // We call this first to ensure we have our filter lists and ad block data ready
+      // Otherwise we sometimes get empty filter lists
+      // We will replace the shield components callback with our `AdblockResourceDownloader`
+      await readyShieldComponents()
+      
+      // Setup ad block resource downloader and start downloading
+      await AdblockResourceDownloader.shared.start(with: braveCore.adblockService)
+      // Setup filter list subscriber and start getting data
+      await FilterListResourceSubscriber.shared.start(with: braveCore.adblockService)
+      
+      // `AdBlockEngineManager.shared.start` should be at the end
+      // as `AdBlockStats` will load an initial engine that will be overriden by
+      // and `AdblockResourceDownloader` and `FilterListResourceSubscriber`
+      // will set some resources on it after calling `start`
+      await AdBlockEngineManager.shared.start()
+      
+      // `ContentBlockerManager.shared.start` should be at the end
+      // as `AdblockResourceDownloader` and `FilterListResourceSubscriber`
+      // will set some resources on it after calling `start`
+      await ContentBlockerManager.shared.start()
+      
+      await didCompileBlockLists()
+    }
+  }
+  
+  private func readyShieldComponents() async {
+    return await withCheckedContinuation { continuation in
+      var ready = false
+      
+      braveCore.adblockService.shieldsComponentReady = { _ in
+        guard !ready else { return }
+        ready = true
+        continuation.resume()
+      }
+    }
+  }
```

Comment:
This isn't a correct usage of a checked continuation because, 1. this callback can be called multiple times and 2. it can be never be called in failure scenarios. 

Set this closure elsewhere and execute your tasks when receiving data or use `AsyncStream`

Reply:
For 1: I make sure it's called only once by setting a boolean. But ill check out AsyncStream.
For 2: I noticed 2 and handled it already.

Reply:
Still isn't a correct usage. This API is essentially just for bridging completion-block APIs into an async context. If you reach for this with a closure that can execute any number of times other than exactly 1 then it isn't the right tool regardless of whether or not you're adding your own checks.

Reply:
Yah but the other tool isn't for this scenario either. And when I use it I get sonar cloud complaining that I use a for loop and take only one result out of it. And I ensure that I only handle one callback completion no more. I also replace the completion block. To be safer I even cleared the completion block right within this callback so its guaranteed to trigger only once.